website/blog/2021-02-04-raptorx.md

```diff
@@ -0,0 +1,214 @@
+
+
+
+![Predicate Pruning](/img/blog/2021-02-04-raptorx/frc.png)
+
+## File Descriptor and Footer Cache
+
+A Presto worker caches the file descriptors in memory to avoid long `openFile` calls to remote storage. Also, a worker caches common columnar file and stripe footers in memory. The current supported file formats are ORC, DWRF, and Parquet. The reason to cache such information in memory is due to the high hit rate of footers as they are the indexes to the data itself.
+
+
+
+## Alluxio Data Cache
+
+Alluxio data cache has been introduced in [an earlier post](https://prestodb.io/blog/2020/06/16/alluxio-datacaching). It is the main feature to deprecate the Raptor connector. A Presto worker caches remote storage data in its original form (compressed and possibly encrypted) on local SSD upon read. If, in the future, there is a read request covering the range that can be found on the local SSD, the request will return the result directly from the local SSD. The caching library was built as a joint effort with [Alluxio](https://www.alluxio.io/) and Presto open source community.
+
+The caching mechanism is to align each read into 1MB chunks, where 1MB is configurable to be adapted to different storage capability. For example, suppose Presto issues a read with 3MB in length starting with offset 0, then Alluxio cache checks if 0 - 1MB, 1 - 2MB, and 2 - 3MB chunks are already on disk and only fetch those that are not cached. The purging policy is based on LRU. It removes chunks from a disk that has not been accessed for the longest time. The Alluxio data cache exposes a standard Hadoop File System interface to the Hive connector, transparently storing requested chunks in a high-performance, highly concurrent, and fault-tolerant storage engine which is designed to serve workloads at Facebook scale.
+
+
+
+![Alluxio Data Cache](/img/blog/2021-02-04-raptorx/alluxio.png)
+
+
+
+
+## Soft Affinity Scheduling
+
+To maximize the cache hit rate on workers, the coordinator needs to schedule the requests of the same file to the same worker. Because there is a high chance part of the file has already been cached on that particular worker. The scheduling policy is "soft", meaning that if the destination worker is too busy or unavailable, the scheduler will fallback to its secondary pick worker for caching or just skip the cache when necessary. The same [earlier post](https://prestodb.io/blog/2020/06/16/alluxio-datacaching) has a detailed explanation of the scheduling policy. The scheduling policy guarantees that cache is not on the critical path but still can boost performance.
```

Comment:
```suggestion
To maximize the cache hit rate on workers, the coordinator needs to schedule the requests of the same file to the same worker. Because there is a high chance part of the file has already been cached on that particular worker. The scheduling policy is "soft", meaning that if the destination worker is too busy or unavailable, the scheduler will fallback to its secondary pick worker for caching or just skip the cache when necessary. The same [earlier post](https://prestodb.io/blog/2020/06/16/alluxio-datacaching) has a detailed explanation of the scheduling policy. The scheduling policy guarantees that cache is not on the critical path, but still can boost performance.
```